### PR TITLE
Add a flag to infer appropriate cross compile hosts on Darwin

### LINKED ISSUE #60348: build-toolchain script fails on M1 Mac with `.../swift-macosx-arm64/bin/swiftc: No such file or directory`
When running `utils/build-toolchain` on an M1 Mac, the build fails when compiling Swift modules with the following error:

```
FAILED: SwiftCompilerSources/_CompilerRegexParser.o /Users/allevato/Source/Swift/build/buildbot_osx/swift-macosx-arm64/SwiftCompilerSources/_CompilerRegexParser.o
cd /Users/allevato/Source/Swift/swift/SwiftCompilerSources && /Users/allevato/Source/Swift/build/buildbot_osx/swift-macosx-arm64/bin/swiftc -c <additional compiler args...>
/bin/sh: /Users/allevato/Source/Swift/build/buildbot_osx/swift-macosx-arm64/bin/swiftc: No such file or directory
```

Indeed, the only binaries in that directory on my machine are the following:

```shell
$ ls /Users/allevato/Source/Swift/build/buildbot_osx/swift-macosx-arm64/bin/
lldb-check-expect lldb-with-tools
```

Additional discussion at https://forums.swift.org/t/get-the-error-when-building-the-toolchain/58665.

### Steps To Reproduce
1. Run `utils/build-toolchain $TOOLCHAIN_ID` (where `TOOLCHAIN_ID` is some arbitrary name)

### Expected behavior
A toolchain is built.

### Environment
 - OS: macOS 12.4
 - Xcode Version/Tag/Branch: Xcode 13.3; apple/swift `main` branch at HEAD

### PR DESCRIPTION
**Pull request by edymtt**
Use that immediately for the macOS package presets.

Addresses rdar://98787335, #60348

### CODE REVIEW

**Comment by edymtt** [comment]
This is based off the suggestion from @shahmishal in #60624

Keeping this as a draft since I expect to iterate on tests and linting.

**Comment by edymtt** [comment]
@swift-ci please python lint

**Review by compnerd — APPROVED** [review]
I'm not sure about the syntax for the macOS test, but if the CI is happy with it, that seems reasonable.  This also addresses my question in the original PR.

File: utils/build_swift/build_swift/driver_arguments.py [hunk 1]
```diff
@@ -580,6 +580,11 @@ def create_argument_parser():
            help='A space separated list of targets to cross-compile host '
                 'Swift tools for. Can be used multiple times.')
 
+    option('--infer-cross-compile-hosts-on-darwin', toggle_true,
+           help="When building on Darwin, populates automatically cross-compile-hosts "
+                "based on the architecture build-script is running on. "
+                "Has precendence over cross-compile-hosts")
```
Comment (edymtt) [inline utils/build_swift/build_swift/driver_arguments.py:586]:
```suggestion
                "Has precedence over cross-compile-hosts")
```

**Comment by edymtt** [comment]
@swift-ci please test

**Comment by edymtt** [comment]
@swift-ci please build toolchain

**Comment by edymtt** [comment]
@swift-ci please build Apple Silicon

**Review by etcwilde — APPROVED** [review]

File: utils/build_swift/build_swift/driver_arguments.py (hunk 1 above)
Comment (edymtt) [reply utils/build_swift/build_swift/driver_arguments.py]: Tackled in c954130

File: utils/build_swift/build_swift/driver_arguments.py [hunk 2]
```diff
@@ -580,6 +580,11 @@ def create_argument_parser():
            help='A space separated list of targets to cross-compile host '
                 'Swift tools for. Can be used multiple times.')
 
+    option('--infer-cross-compile-hosts-on-darwin', toggle_true,
+           help="When building on Darwin, populates automatically cross-compile-hosts "
```
[the remark this answers is not in the document]
Comment (edymtt) [reply utils/build_swift/build_swift/driver_arguments.py]: Thanks -- tackled in c954130

**Review by shahmishal — APPROVED** [review]
Thanks for changing this to dynamic.

**Comment by edymtt** [comment]
* fixed `validation-test/BuildSystem/infer-cross-compile-hosts-on-darwin-macosx.test` to rely instead on `%target-cpu` -- the `%if` feature I was using is not present in `stable/20220421` (https://reviews.llvm.org/rG1041a9642ba035fd2685f925911d705e8edf5bb0)
* split the usage of the new flag in a separate commit, so in case of regressions that can be reverted separately while leaving the bulk of the logic in place (but inert)

**Comment by edymtt** [comment]
@swift-ci please test

**Comment by edymtt** [comment]
@swift-ci please build toolchain

**Comment by edymtt** [comment]
@swift-ci please build Apple Silicon

**Comment by edymtt** [comment]
@swift-ci please python lint

**Review by drexin — APPROVED** [review]
LGTM, thanks for doing this 🙂

**Comment by edymtt** [comment]
Hitting an unrelated failure in macOS PR testing now solved by #60709, retrying

@swift-ci please macOS

**Comment by rayxia** [comment]
@edymtt Hi, after using this code, I got another error when buidling the toolchain, can you help me?

```
Undefined symbols for architecture arm64:
  "llvm::cl::Option::anchor()", referenced from:
      vtable for llvm::cl::list<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool, llvm::cl::parser<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > in swift-demangle.cpp.o
  "llvm::cl::Option::addOccurrence(unsigned int, llvm::StringRef, llvm::StringRef, bool)", referenced from:
      vtable for llvm::cl::list<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool, llvm::cl::parser<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > in swift-demangle.cpp.o
  "llvm::cl::GenericOptionValue::anchor()", referenced from:
      vtable for llvm::cl::OptionValueCopy<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > in swift-demangle.cpp.o
      vtable for llvm::cl::OptionValue<bool> in swift-demangle.cpp.o
  "vtable for llvm::cl::parser<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >", referenced from:
      __GLOBAL__sub_I_swift_demangle.cpp in swift-demangle.cpp.o
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
  "llvm::cl::Option::addArgument()", referenced from:
      __GLOBAL__sub_I_swift_demangle.cpp in swift-demangle.cpp.o
  "vtable for llvm::cl::OptionValue<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >", referenced from:
      __GLOBAL__sub_I_swift_demangle.cpp in swift-demangle.cpp.o
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
  "llvm::cl::Option::setArgStr(llvm::StringRef)", referenced from:
      __GLOBAL__sub_I_swift_demangle.cpp in swift-demangle.cpp.o
  "llvm::SmallVectorBase<unsigned int>::grow_pod(void*, unsigned long, unsigned long)", referenced from:
      llvm::cl::Option::Option(llvm::cl::NumOccurrencesFlag, llvm::cl::OptionHidden) in swift-demangle.cpp.o
  "vtable for llvm::cl::opt<bool, false, llvm::cl::parser<bool> >", referenced from:
      llvm::cl::opt<bool, false, llvm::cl::parser<bool> >::~opt() in swift-demangle.cpp.o
      __GLOBAL__sub_I_swift_demangle.cpp in swift-demangle.cpp.o
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
  "llvm::cl::basic_parser_impl::getOptionWidth(llvm::cl::Option const&) const", referenced from:
      llvm::cl::list<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool, llvm::cl::parser<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >::getOptionWidth() const in swift-demangle.cpp.o
  "llvm::cl::ParseCommandLineOptions(int, char const* const*, llvm::StringRef, llvm::raw_ostream*, char const*, bool)", referenced from:
      _main in swift-demangle.cpp.o
  "llvm::raw_ostream::write(unsigned char)", referenced from:
      _main in swift-demangle.cpp.o
      demangle(llvm::raw_ostream&, llvm::StringRef, swift::Demangle::Context&, swift::Demangle::DemangleOptions const&) in swift-demangle.cpp.o
      llvm::raw_ostream::operator<<(char) in swift-demangle.cpp.o
  "vtable for llvm::cl::parser<bool>", referenced from:
      __GLOBAL__sub_I_swift_demangle.cpp in swift-demangle.cpp.o
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
  "llvm::cl::getGeneralCategory()", referenced from:
      llvm::cl::Option::Option(llvm::cl::NumOccurrencesFlag, llvm::cl::OptionHidden) in swift-demangle.cpp.o
  "llvm::cl::basic_parser_impl::printOptionInfo(llvm::cl::Option const&, unsigned long) const", referenced from:
      llvm::cl::list<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool, llvm::cl::parser<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >::printOptionInfo(unsigned long) const in swift-demangle.cpp.o
  "llvm::outs()", referenced from:
      _main in swift-demangle.cpp.o
      demangle(llvm::raw_ostream&, llvm::StringRef, swift::Demangle::Context&, swift::Demangle::DemangleOptions const&) in swift-demangle.cpp.o
  "llvm::errs()", referenced from:
      demangle(llvm::raw_ostream&, llvm::StringRef, swift::Demangle::Context&, swift::Demangle::DemangleOptions const&) in swift-demangle.cpp.o
  "llvm::raw_ostream::write(char const*, unsigned long)", referenced from:
      _main in swift-demangle.cpp.o
      demangle(llvm::raw_ostream&, llvm::StringRef, swift::Demangle::Context&, swift::Demangle::DemangleOptions const&) in swift-demangle.cpp.o
      llvm::raw_ostream::operator<<(llvm::StringRef) in swift-demangle.cpp.o
      llvm::raw_ostream::operator<<(char const*) in swift-demangle.cpp.o
      llvm::raw_ostream::operator<<(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in swift-demangle.cpp.o
  "vtable for llvm::cl::opt<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, false, llvm::cl::parser<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >", referenced from:
      llvm::cl::opt<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, false, llvm::cl::parser<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >::~opt() in swift-demangle.cpp.o
      __GLOBAL__sub_I_swift_demangle.cpp in swift-demangle.cpp.o
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
  "llvm::raw_ostream::operator<<(long)", referenced from:
      llvm::raw_ostream::operator<<(int) in swift-demangle.cpp.o
  "llvm::raw_ostream::operator<<(unsigned long)", referenced from:
      llvm::raw_ostream::operator<<(unsigned int) in swift-demangle.cpp.o
  "vtable for llvm::cl::Option", referenced from:
      llvm::cl::opt<bool, false, llvm::cl::parser<bool> >::~opt() in swift-demangle.cpp.o
      llvm::cl::opt<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, false, llvm::cl::parser<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >::~opt() in swift-demangle.cpp.o
      llvm::cl::list<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool, llvm::cl::parser<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >::~list() in swift-demangle.cpp.o
      llvm::cl::Option::Option(llvm::cl::NumOccurrencesFlag, llvm::cl::OptionHidden) in swift-demangle.cpp.o
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
ld: symbol(s) not found for architecture arm64
```

**Comment by edymtt** [comment]
@rayxia this is not something I hit so far when building on Apple Silicon

I would suggest creating an issue at https://github.com/apple/swift/issues/new/choose
providing at least
* the machine you are building on (OS and architecture)
* the build-script/build-toolchain invocation you are using

If possible it would be nice to have
* the full log from the run
* the build.ninja files from any build/*/swift-macosx-<arch> folders you have

(not sure how much I can follow up since I'm busy with other efforts, but this way you others may be more likely to see your issue)

**Comment by rayxia** [comment]
> @rayxia this is not something I hit so far when building on Apple Silicon
> 
> I would suggest creating an issue at https://github.com/apple/swift/issues/new/choose providing at least
> 
> * the machine you are building on (OS and architecture)
> * the build-script/build-toolchain invocation you are using
> 
> If possible it would be nice to have
> 
> * the full log from the run
> * the build.ninja files from any build/*/swift-macosx- folders you have
> 
> (not sure how much I can follow up since I'm busy with other efforts, but this way you others may be more likely to see your issue)

@edymtt OK, thanks for your suggestion.

**Pull request merged**